### ISSUE TITLE
Do not move output A windows if output B is added or removed

### ISSUE THREAD

**Issue by Tamaranch** [labwc/labwc#1541]
This is an enhancement request. Currently (0.7.0, git-master), when windows are opened on an output A and another output B is added or removed (enabled/disabled via wlr-output-management), the windows of output A are moved, by a certain distance probably proportional to the size of output B.

I'd expect the windows in output A not to move at all, since their position in output A doesn't depend on the presence or absence of other outputs. This is what happens in Wayfire 0.8.0, for example.

Thanks!

**Comment by Consolatis** [comment]
Is the new output maybe added at `0,0`?
Otherwise this should only happen when a window on output A is partly out of its usable area (e.g. output size - exclusive layershell clients like panels).

**Comment by Tamaranch** [comment]
> Is the new output maybe added at `0,0`?

Yes indeed, laptop screen on the left, external screen on the right, and I remove or put back the laptop screen.

**Comment by ahesford** [comment]
For now, if you put the changing out at negative coordinates, the windows should not move. The problem is that you've moved the external output relative to the fixed virtual coordinate space used for placing and sizing objects. Effectively, the world has stayed the same, but you've moved the "camera" relative to the world. By keeping the external display at (0, 0) and putting the laptop display at a negative position, you keep the external "camera" fixed.

Tacking windows to outputs so that they move when the output coordinates change would require a bit more bookkeeping, but would be possible.

**Comment by johanmalm** [comment]
Could this be achieved with kanshi by defining two profiles - one for single output setup at (X,Y) and another for the two-output setup at (0,0) and (X,Y). I can try over the weekend if helpful.

**Comment by ahesford** [comment]
I believe a kanshi profile along the lines of

```
profile laptop {
    output eDP-1 enable position 0, 0
}

profile external-only {
    output eDP-1 disable
    output HDMI-A-1 enable position 0, 0
}

profile external-plus {
    output eDP-1 enable position -1920, 0
    output HDMI-A-1 enable position 0, 0
}
```

should provide a means to switch between external-only and external-plus-internal without reconfiguring all views on your external display. If you turn off the internal display with `kanshictl switch external-only`, all the windows on the internal display will be gathered on the external display. As long as you don't remove those windows, `kanshictl switch external-plus` will activate the laptop display and bring those windows back to the laptop. When you yank the external display, it will re-activate the internal display at (0, 0) and gather everything.

**Comment by Tamaranch** [comment]
Thanks for the suggestion with kanshi, but I have the impression that it appears to be a competing tool in my case, as I end up with these configurations via xfce4-display-settings, so I don't want to change them afterwards (including the output coordinates in compositor space).

> --EDIT-- Or do you mean that we should leave windows where they are rather than moving them to the remaining output?

Well, I don't really understand why they are moved to the remaining output in this case, since they are already there. I think what I'm expecting is that if all or part of a window is on an output that remains in place after a configuration change (regardless of its position in compositor space), then it doesn't move. On the other hand, if a window is no longer on any output, it is obviously moved to the nearest output.

**Comment by Consolatis** [comment]
> Well, I don't really understand why they are moved to the remaining output in this case, since they are already there. 

The windows keep their layout position, when only the external (right) monitor is part of the layout all window shown there are aligned to the top left corner (0,0) of the layout and thus to the top left corner of the external monitor. When a new monitor is added to the layout at 0,0 and the external one is moved to something like 1920,0 the windows are still on the same layout position but that position is now on the left monitor. So we don't actually move the windows at all in this case, its just that the monitor viewports to the layout change.

**Comment by Tamaranch** [comment]
> So we don't actually move the windows at all in this case, its just that the monitor viewports to the layout change.

Yes I understand that, but from the user's point of view the windows move :) My request would probably be to work in the output coordinate system rather than the compositor coordinate system. I imagine that's what wayfire does, at least in this case.

Anyway, if that's how it's going to stay, that's fine by me too ^^'.

**Comment by Consolatis** [comment]
Does it also happen when both outputs are enabled (internal left, external right) and then the internal one gets disabled and later re-enabled? If yes, what is the position of the external monitor  (via `wlr-randr` to eliminate any potential stale state) after disabling the internal one?

**Comment by Tamaranch** [comment]
> Does it also happen when both outputs are enabled (internal left, external right) and then the internal one gets disabled and later re-enabled?

Yes, this is my use case.

> Maybe we could enforce it being whatever its old position was instead of letting automatic output layout move it to 0,0. 

This is not an automatic move, labwc does what xfce4-display-settings tells it to via `zwlr_output_configuration_head_v1_set_position()`.

**Comment by Consolatis** [comment]
I think we may indirectly change the output position of the external monitor (which should be at `1920,0` or something like that when it was the right monitor before). That is unless xfce4-display-settings sets it to `0,0` itself.

We do use a wlroots feature for automatically adjusting the output positions within the layout based on the other outputs. That will be overwritten whenever a client sets its own position that differs from the current position within the layout for a given output. So what might happen here is that we don't set the output to the fixed position (because its the same as currently configured) and then remove the 2nd output from the layout which in turn lets the automatic layout kick in and moves the external monitor to `0,0` even if it was configured by the client to use its earlier position.
